### PR TITLE
feat(atomic): re-render localized string when language changes

### PR DESCRIPTION
**Pull request by ThibodeauJF**
- Renamed `context` to `bindings` and made it public, since it technically is being modified by the SearchInterface
- Added the public `updateLocaleStrings` method which, when defined, is called atomically on init & language change
https://coveord.atlassian.net/browse/KIT-348

### CODE REVIEW

File: packages/atomic/src/components/atomic-search-box/atomic-search-box.tsx [hunk 1]
```diff
@@ -46,14 +46,19 @@ export class AtomicSearchBox implements AtomicSearchBoxOptions {
...
-  private context!: InterfaceContext;
+  public context!: InterfaceContext;
   private searchBox!: SearchBox;
   private unsubscribe: Unsubscribe = () => {};
   private inputRef!: HTMLInputElement;
   private valuesRef!: HTMLElement;
   private containerRef!: HTMLElement;
   private combobox!: Combobox;
 
+  @State() strings = {
```
Comment (olamothe) [inline packages/atomic/src/components/atomic-search-box/atomic-search-box.tsx:57]: As discussed this morning, it would be awesome if we had something to abstract the concepts of localized string + how to handle the `updateLocaleStrings`.

A decorator, like you proposed would be nice.

So that a newcomer can simply copy paste the recipe to add new localized string, or to create a new component with localized string, and it's super straightforward to do.

**Review by olamothe — APPROVED** [review]

File: packages/atomic/src/components/atomic-search-box/atomic-search-box.tsx (hunk 1 above)
Comment (ThibodeauJF) [reply packages/atomic/src/components/atomic-search-box/atomic-search-box.tsx:57]: Hmm that's what I do to the extent I can in the Initialization decorator, it calls the `updateLocaleStrings` directly, only the user has to do is define the update method to update the state.

To have the right types, we need to define a state on line 57, but we can't access the context property yet. I can look a bit further

Comment (ThibodeauJF) [reply packages/atomic/src/components/atomic-search-box/atomic-search-box.tsx]: @olamothe 
Okay I now added methods directly in the state, the state is then automatically updated when the language state (in the Initialization decorator of the initialization-utils.tsx file, see latest commit)

File: packages/atomic/src/utils/initialization-utils.tsx [hunk 2]
```diff
@@ -19,37 +19,47 @@ export class InitializationError extends Error {
...
       const element = getElement(this);
       const event = new CustomEvent('atomic/initializeComponent', {
-        detail: (context: InterfaceContext) => {
-          this[contextProperty] = context;
+        detail: (bindings: Bindings) => {
+          this.bindings = bindings;
+          if (this.strings) {
+            this.bindings.i18n.on(
+              'languageChanged',
+              () => (this.strings = {...this.strings})
```
Comment (ThibodeauJF) [inline packages/atomic/src/utils/initialization-utils.tsx:46]: This re-renders the component correctly

File: packages/atomic/src/utils/initialization-utils.tsx [hunk 3]
```diff
@@ -19,37 +19,55 @@ export class InitializationError extends Error {
...
       const element = getElement(this);
       const event = new CustomEvent('atomic/initializeComponent', {
-        detail: (context: InterfaceContext) => {
-          this[contextProperty] = context;
+        detail: (bindings: Bindings) => {
+          this.bindings = bindings;
+
+          // Ensures re-render of localized strings
+          if (this.strings) {
+            this.strings = {...this.strings};
```
Comment (samisayegh) [inline packages/atomic/src/utils/initialization-utils.tsx:51]: Why is this line needed? It seems like an unnecessary rerender?

File: packages/atomic/src/utils/initialization-utils.tsx [hunk 4]
```diff
@@ -19,37 +19,55 @@ export class InitializationError extends Error {
...
+            this.bindings.i18n.on(
```
Comment (samisayegh) [inline packages/atomic/src/utils/initialization-utils.tsx:52]: Should we unsubscribe this event listener when a component unmounts to avoid memory leaks?

File: packages/atomic/README.md [hunk 5]
```diff
@@ -81,3 +81,62 @@ To run all the test, run:
...
+  private unsubscribe: Unsubscribe = () => {};
+
+  @Initialization()
+  public initialize() {
+    this.controller = buildController(this.bindings.engine);
+    this.unsubscribe = this.controller.subscribe(() => {
+      this.state = this.controller.state;
+    });
+  }
+
```
Comment (samisayegh) [inline packages/atomic/README.md:137]: I think we should include the `unmount` life-cycle in the example to remind developers to call `unsubscribe`.

**Review by samisayegh — APPROVED** [review]

File: packages/atomic/src/utils/initialization-utils.tsx (hunk 3 above)
Comment (ThibodeauJF) [reply packages/atomic/src/utils/initialization-utils.tsx:51]: It's not, component needs to render the new string on initialization, without it, we see undefined

File: packages/atomic/README.md (hunk 5 above)
Comment (ThibodeauJF) [reply packages/atomic/README.md:137]: A component with bindings won't necessarily need an unsubscribe if they're not using a controller/engine to subscribe to

Comment (ThibodeauJF) [reply packages/atomic/README.md:137]: Maybe if we set the "controller" property to an AtomicComponentInterface, and maybe even the updateState, we could check if it's defined and then call unsubscribe 🤔

**Comment by ThibodeauJF** [comment]
Refactoring quite a few things... will open a fresh PR with more info. Sorry!
